File: api/TRPG.Contracts/Combat/Responses/CombatRoundEvent.cs
using System.Text.Json.Serialization;

namespace TRPG.Contracts.Combat.Responses;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(CombatHitEvent), nameof(CombatHitEvent))]
[JsonDerivedType(typeof(CombatMissEvent), nameof(CombatMissEvent))]
[JsonDerivedType(typeof(CombatBlockEvent), nameof(CombatBlockEvent))]
[JsonDerivedType(typeof(CombatRegeneratedEvent), nameof(CombatRegeneratedEvent))]
[JsonDerivedType(typeof(CombatResourceStateUpdatedEvent), nameof(CombatResourceStateUpdatedEvent))]
public abstract record CombatRoundEvent(
    Guid AttackerId,
    string AttackerName,
    string AbilityName,
    Guid TargetId,
    string TargetName
);

public sealed record CombatHitEvent(
    Guid AttackerId,
    string AttackerName,
    string AbilityName,
    Guid TargetId,
    string TargetName,
    int Damage,
    DamageType DamageType,
    bool IsCritical,
    bool Killed,
    int TargetRemainingHp,
    int TargetMaximumHp,
    IReadOnlyList<ConditionType> AppliedConditions
) : CombatRoundEvent(AttackerId, AttackerName, AbilityName, TargetId, TargetName);

public sealed record CombatMissEvent(
    Guid AttackerId,
    string AttackerName,
    string AbilityName,
    Guid TargetId,
    string TargetName
) : CombatRoundEvent(AttackerId, AttackerName, AbilityName, TargetId, TargetName);

public sealed record CombatBlockEvent(
    Guid AttackerId,
    string AttackerName,
    string AbilityName,
    Guid TargetId,
    string TargetName
) : CombatRoundEvent(AttackerId, AttackerName, AbilityName, TargetId, TargetName);

public sealed record CombatRegeneratedEvent(
    Guid AttackerId,
    string AttackerName,
    int PreviousAp,
    int CurrentAp,
    int MaximumAp,
    int PreviousMp,
    int CurrentMp,
    int MaximumMp
) : CombatRoundEvent(AttackerId, AttackerName, "Regenerate", AttackerId, AttackerName);

public sealed record CombatResourceStateUpdatedEvent(
    Guid AttackerId,
    string AttackerName,
    int CurrentAp,
    int MaximumAp,
    int CurrentMp,
    int MaximumMp
) : CombatRoundEvent(AttackerId, AttackerName, "Resource state updated", AttackerId, AttackerName);

File: spa/src/features/combat/combat-round-event.ts
import type {
  CombatUpdatePayload as GeneratedCombatUpdatePayload,
  DamageType,
} from '@/api/signalr-client/TRPG.Combat.ClientModels';

export type CombatActionEvent = CombatHitEvent | CombatMissEvent | CombatBlockEvent;

export type CombatRoundEvent =
  | CombatActionEvent
  | CombatRegeneratedEvent
  | CombatResourceStateUpdatedEvent;

interface CombatRoundEventBase {
  attackerId: string;
  attackerName: string;
  abilityName: string;
  targetId: string;
  targetName: string;
  narration?: string | null;
}

export interface CombatHitEvent extends CombatRoundEventBase {
  type: 'CombatHitEvent';
  damage: number;
  damageType: DamageType;
  isCritical: boolean;
  killed: boolean;
  targetRemainingHp: number;
  targetMaximumHp: number;
  appliedConditions: string[];
}

export interface CombatMissEvent extends CombatRoundEventBase {
  type: 'CombatMissEvent';
}

export interface CombatBlockEvent extends CombatRoundEventBase {
  type: 'CombatBlockEvent';
}

export interface CombatRegeneratedEvent {
  type: 'CombatRegeneratedEvent';
  attackerId: string;
  attackerName: string;
  previousAp: number;
  currentAp: number;
  maximumAp: number;
  previousMp: number;
  currentMp: number;
  maximumMp: number;
}

export interface CombatResourceStateUpdatedEvent {
  type: 'CombatResourceStateUpdatedEvent';
  attackerId: string;
  attackerName: string;
  currentAp: number;
  maximumAp: number;
  currentMp: number;
  maximumMp: number;
}

export type CombatUpdatePayload = Pick<GeneratedCombatUpdatePayload, 'combatants' | 'outcome'> & {
  events: CombatRoundEvent[];
};

export function toCombatUpdatePayload(payload: GeneratedCombatUpdatePayload): CombatUpdatePayload {
  return { ...payload, events: payload.events as CombatRoundEvent[] };
}
